### PR TITLE
Unit testing task

### PR DESCRIPTION
**Pull request by VladSelikh**

### CODE REVIEW

File: test/specs/calculator.add.multiply.spec.js [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+  it('It should test that if \'number\' is null an error is thrown', () => {
+    expect(() => calculator.add(null, 5, 6)()).to.throw('Unable to calculate, [null] is not a number');
+  });
+  // eslint-disable-next-line max-len
+  it('It should test that if \'increment\' is not defined(null) an error is thrown', () => {
+    expect(() => calculator.add(68, null)()).to.throw('Unable to calculate, [null] is not a number');
+  });
+  // eslint-disable-next-line max-len
+  it('It should test that if \'number\' and \'increment\' are not defined(null) an error is thrown', () => {
+    expect(() => calculator.add(null, null)()).to.throw('Unable to calculate, [null] is not a number');
```
Comment (VladSelikh) [inline test/specs/calculator.add.multiply.spec.js]: It's a bad practice to test a scenario with 2 invalid arguments. Please correct.

File: test/specs/calculator.add.multiply.spec.js [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  it('It should test that if only \'number\' is defined(null) an error is thrown', () => {
+    expect(() => calculator.add(null)()).to.throw('Unable to calculate, [null] is not a number');
+  });
+  // eslint-disable-next-line max-len
+  it('It should test that if only \'number\' is defined(as undefined) an error is thrown', () => {
+    expect(() => calculator.add(undefined)()).to.throw('Unable to calculate, [undefined] is not a number');
+  });
+  // eslint-disable-next-line max-len
+  it('It should test that if \'number\' and \'increment\' are not defined(undefined) an error is thrown', () => {
+    expect(() => calculator.add(undefined, undefined)()).to.throw('Unable to calculate, [undefined] is not a number');
```
Comment (VladSelikh) [inline test/specs/calculator.add.multiply.spec.js]: It's a bad practice to test a scenario with 2 invalid arguments. Please correct.

File: test/specs/calculator.add.multiply.spec.js [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+  afterEach(() => {
+    multiplier = null;
+  });
+  // eslint-disable-next-line max-len
+  it('It should test that if \'number\' is not defined(null) an error is thrown', () => {
+    expect(() => multiplier.multiply(68, null)()).to.throw('Unable to multiply, [null] is not a number');
+  });
+  // eslint-disable-next-line max-len
+  it('It should test that if \'increment\' is not defined(null) an error is thrown', () => {
+    expect(() => multiplier.multiply(68, null)()).to.throw('Unable to multiply, [null] is not a number');
```
Comment (VladSelikh) [inline test/specs/calculator.add.multiply.spec.js]: Is there any difference between this test and the previous one?

File: test/specs/calculator.add.multiply.spec.js [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+  });
+  // eslint-disable-next-line max-len
+  it('It should test that if \'number\' and \'multiplier\' are not defined(null) an error is thrown', () => {
+    expect(() => multiplier.multiply(null, null)()).to.throw('Unable to multiply, [null] is not a number');
```
Comment (VladSelikh) [inline test/specs/calculator.add.multiply.spec.js]: It's a bad practice to test a scenario with 2 invalid arguments. Please correct.

File: test/specs/calculator.add.multiply.spec.js [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+  it('It should test that if the only \'number\' is defined(null) an error is thrown', () => {
+    expect(() => multiplier.multiply(null)()).to.throw('Unable to multiply, [null] is not a number');
+  });
+  // eslint-disable-next-line max-len
+  it('It should test that if the only \'number\' is defined(as undefined) an error is thrown', () => {
+    expect(() => multiplier.multiply(undefined)()).to.throw('Unable to multiply, [undefined] is not a number');
+  });
+  // eslint-disable-next-line max-len
+  it('It should test that if \'number\' and \'multiplier\' are not defined(undefined) an error is thrown', () => {
+    expect(() => multiplier.multiply(undefined, undefined)()).to.throw('Unable to multiply, [undefined] is not a number');
```
Comment (VladSelikh) [inline test/specs/calculator.add.multiply.spec.js]: It's a bad practice to test a scenario with 2 invalid arguments. Please correct.

File: test/specs/calculator.add.multiply.spec.js [hunk 6]
```diff
@@ -0,0 +1,158 @@
...
+  });
+  // eslint-disable-next-line max-len
+  it('It should test that if \'number\' is not defined(undefined) an error is thrown', () => {
+    expect(() => multiplier.multiply(undefined, 5)()).to.throw('Unable to multiply, [undefined] is not a number');
+  });
+  // eslint-disable-next-line max-len
+  it('It should test that if \'increment\' is not defined(undefined) an error is thrown', () => {
+    expect(() => multiplier.multiply(68, undefined)()).to.throw('Unable to multiply, [undefined] is not a number');
+  });
```
Comment (VladSelikh) [inline test/specs/calculator.add.multiply.spec.js:152]: Please add negative tests with other types of arguments: `string`, `object`, etc.

File: app/calculator.js [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+          } else if (typeof (current) !== 'number') {
+            // eslint-disable-next-line max-len
+            throw new TypeError(`Unable to calculate, [${current}] is not a number`);
+          }
+        } else {
+          return sum + current;
+        }
+      }, 0);
+    }
+  }
```
Comment (VladSelikh) [inline app/calculator.js:27]: You've implemented too complicated logic here.
```
add(...numbers) {
    if (numbers.some((number) => typeof number !== 'number')) {
      const invalidArgument = numbers.find((arg) => typeof arg !== 'number');
      throw new Error(`Provided value ${invalidArgument} is not in the list of allowed data types`);
    } else {
      return numbers.reduce((previous, current) => previous + current);
    }
  }
  ```
  This variant is much better. The same is for `multiply()` method.

File: test/specs/calculator.add.multiply.spec.js (hunk 1 above)
Comment (axferox) [reply test/specs/calculator.add.multiply.spec.js]: Why? Due to the fact that our methods can receive any amount of any values I don't think it's an invalid check.

File: test/specs/calculator.add.multiply.spec.js (hunk 5 above)
Comment (axferox) [reply test/specs/calculator.add.multiply.spec.js]: Why? Due to the fact that our methods can receive any amount of any values I don't think it's an invalid check.

File: test/specs/calculator.add.multiply.spec.js (hunk 4 above)
Comment (axferox) [reply test/specs/calculator.add.multiply.spec.js]: Why? Due to the fact that our methods can receive any amount of any values I don't think it's an invalid check.

File: test/specs/calculator.add.multiply.spec.js (hunk 2 above)
Comment (axferox) [reply test/specs/calculator.add.multiply.spec.js]: Why? Due to the fact that our methods can receive any amount of any values I don't think it's an invalid check.

File: app/calculator.js (hunk 7 above)
Comment (axferox) [reply app/calculator.js:27]: This code does not work for cases when we ain't put nothing into a function or putting only one number:
In the first case we're getting: 
'Reduce of empty array with no initial value'
In second case we're getting: 
'-calculator.add(...) is not a function'

That's a reason why the complicated logic was implemented, you need to 'refine' the values before putting them into a reduce

File: test/specs/calculator.add.multiply.spec.js (hunk 5 above)
Comment (VladSelikh) [reply test/specs/calculator.add.multiply.spec.js]: In this case, can not be sure of the exact reason for your test being failed.

File: test/specs/calculator.add.multiply.spec.js (hunk 4 above)
Comment (VladSelikh) [reply test/specs/calculator.add.multiply.spec.js]: In this case, can not be sure of the exact reason for your test being failed.

File: test/specs/calculator.add.multiply.spec.js (hunk 2 above)
Comment (VladSelikh) [reply test/specs/calculator.add.multiply.spec.js]: In this case, can not be sure of the exact reason for your test being failed.

File: test/specs/calculator.add.multiply.spec.js (hunk 1 above)
Comment (VladSelikh) [reply test/specs/calculator.add.multiply.spec.js]: In this case, can not be sure of the exact reason for your test being failed.

**Pull request merged**